plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java

```diff
@@ -255,4 +279,21 @@ public static Optional<HiveColumnHandle> toHiveColumnHandle(DeltaLakeColumnHandl
                 throw new IllegalArgumentException("Unsupported column mapping: " + columnMapping);
         }
     }
+
+    private static ConnectorPageSource createPages(long totalRowCount)
+    {
+        // This implementation of the page source is used whenever a query doesn't reference any data
+        // columns. We need to limit the number of rows per page in case there are projections
+        // in the query that can cause page sizes to explode. For example: SELECT rand() FROM some_table
+        int maxPageSize = 100_000;
+
+        ImmutableList.Builder<Page> pages = ImmutableList.builder();
+        for (long remaining = totalRowCount; remaining > 0; remaining -= maxPageSize) {
+            int pageSize = toIntExact(min(maxPageSize, totalRowCount - remaining));
+            pages.add(new Page(pageSize));
```

Comment:
I may be very wrong about this, why do you produce empty pages here? what is the point ?

Reply:
`computeActual("SELECT count(*) FROM test_read_part_key WHERE key = 'p1'")` returns `0` in `io.trino.plugin.deltalake.TestDeltaLakeFileOperations#testReadWholePartition` 

Good catch @homar 

@findepi  It is worth adding assertions for the selections in the `TestDeltaLakeFileOperations`

Reply:
this is plugged into `DeltaLakePageSource`.  that class will project (add) partitioning  columns, but it needs to know how many rows to project to. typically this happens by reading the parquet file (`ParquetPageSourceFactory.createPageSource`), but in this special case this is not necessary.

(other option would be to modify the `ParquetPageSourceFactory.createPageSource` and special case there; thoughts? cc @raunaqmorarka @losipiuk )

Reply:
Since this optimization is only possible by knowing row count from file statistics of delta, I would keep this code right here in delta.

Reply:
> `computeActual("SELECT count(*) FROM test_read_part_key WHERE key = 'p1'")` returns `0` in `io.trino.plugin.deltalake.TestDeltaLakeFileOperations#testReadWholePartition`

indeed, there is a bug on this line
```
int pageSize = toIntExact(min(maxPageSize, totalRowCount - remaining));
```

should be

```
int pageSize = toIntExact(min(maxPageSize, remaining));
```

> It is worth adding assertions for the selections in the TestDeltaLakeFileOperations

maybe, but note that many many tests failed
(in the earlier version i had this differently, with iterator, that's why i didn't notice)